commons/src/main/java/io/aiven/kafka/tieredstorage/commons/storage/s3/S3OutputStream.java

```diff
@@ -0,0 +1,156 @@
+    public static final int DEFAULT_PART_SIZE = 5 * 1024 * 1024;
+
+    private final AmazonS3 client;
+    private final ByteBuffer partBuffer;
+    private final String bucketName;
+    private final String key;
+
+    private MultipartUpload multipartUpload;
+
+    private final int partSize;
+
+    public S3OutputStream(final String bucketName,
+                          final String key,
+                          final int partSize,
+                          final AmazonS3 client) {
+        this.bucketName = bucketName;
+        this.key = key;
+        this.client = client;
+        this.partSize = partSize;
+        this.partBuffer = ByteBuffer.allocate(partSize);
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+        write(new byte[] {(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(final byte[] b,
+                      final int off,
+                      final int len) throws IOException {
+        if (b.length == 0) {
+            return;
+        }
+        if (Objects.isNull(multipartUpload)) {
+            multipartUpload = newMultipartUpload();
+        }
+        final ByteBuffer source = ByteBuffer.wrap(b, off, len);
+        while (source.hasRemaining()) {
+            final int transferred = Math.min(partBuffer.remaining(), source.remaining());
+            final int offset = source.arrayOffset() + source.position();
+            partBuffer.put(source.array(), offset, transferred);
+            source.position(source.position() + transferred);
+            if (!partBuffer.hasRemaining()) {
+                flushBuffer(0, partSize, partSize);
+            }
+        }
+    }
+
+    private MultipartUpload newMultipartUpload() {
+        log.debug("Create new multipart upload request");
+        final var initialRequest = new InitiateMultipartUploadRequest(bucketName, key);
+        final var initiateResult = client.initiateMultipartUpload(initialRequest);
+        log.debug("Upload ID: {}", initiateResult.getUploadId());
+        return new MultipartUpload(initiateResult.getUploadId());
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (partBuffer.position() > 0 && Objects.nonNull(multipartUpload)) {
+            flushBuffer(partBuffer.arrayOffset(), partBuffer.position(), partBuffer.position());
+        }
+        if (Objects.nonNull(multipartUpload)) {
+            multipartUpload.complete();
+        }
+    }
+
+    private void flushBuffer(final int offset,
+                             final int length,
+                             final int partSize) throws IOException {
+        try {
+            multipartUpload.uploadPart(
+                new ByteArrayInputStream(partBuffer.array(), offset, length),
+                partSize
+            );
+            partBuffer.clear();
+        } catch (final Exception e) {
+            multipartUpload.abort();
```

Comment:
Note: I made a ticket for us to monitor/auto-abort unaborted uploads, which may happen in case of a process failure or killing